// dockerlint/lib/parser.js
// Generated by CoffeeScript 1.12.4
(function() {
  var fs, utils;

  fs = require('fs');

  utils = require(__dirname + "/utils");

  exports.getInstruction = function(s) {
    var instruction;
    instruction = s.split(/[ \t]/)[0];
    if (instruction.beginsWith('#')) {
      return 'comment';
    } else {
      return instruction;
    }
  };

  exports.getArguments = function(s) {
    var inst;
    inst = this.getInstruction(s);
    if (inst === 'comment') {
      inst = '#';
    }
    return [s.replace(inst, '').replace(/\\(\s*)$/, '').trim()];
  };

  exports.parser = function(dockerfile) {
    var rules, self;
    rules = [];
    self = this;
    (function() {
      var cont, data, e, i, len, line, lineno, results, rule;
      lineno = 1;
      cont = false;
      rule = [];
      try {
        data = fs.readFileSync(dockerfile).toString().split(/\r?\n/);
      } catch (error) {
        e = error;
        return [];
      }
      results = [];
      for (i = 0, len = data.length; i < len; i++) {
        line = data[i];
        if (utils.notEmpty(line) && !line.beginsWith('#')) {
          if (line.endsWith('\\')) {
            if (cont) {
              rule[0]["arguments"] = rule[0]["arguments"].concat(self.getArguments(line));
            } else {
              cont = true;
              rule.push({
                raw: line,
                line: lineno,
                instruction: self.getInstruction(line),
                "arguments": self.getArguments(line)
              });
            }
          } else if (cont && !line.endsWith('\\')) {
            rules.push({
              raw: line,
              line: rule[0].line,
              instruction: rule[0].instruction,
              "arguments": rule[0]["arguments"].concat(self.getArguments(line))
            });
            rule = [];
            cont = false;
          } else if (!(line.endsWith('\\' && cont))) {
            rules.push({
              raw: line,
              line: lineno,
              instruction: self.getInstruction(line),
              "arguments": self.getArguments(line)
            });
          }
        } else if (line.beginsWith('#')) {
          rules.push({
            raw: line,
            line: lineno,
            instruction: self.getInstruction(line),
            "arguments": self.getArguments(line)
          });
        }
        results.push(lineno++);
      }
      return results;
    })();
    return rules;
  };

}).call(this);
